src/handlers/contributor.rs

```diff
@@ -50,6 +100,104 @@ impl Contributor {
+                    // Get contributor
+                    let Some(contributor) = ordered_contributors.get(&s) else {
+                        info!("contributor not found: {:?}", s);
+                        continue;
+                    };
+
+                    // Check if contributor already signed
+                    let Some(signatures) = signatures.get_mut(&round) else {
+                        info!("signatures not found: {:?}", round);
+                        continue;
+                    };
+                    if signatures.contains_key(contributor) {
+                        info!("contributor already signed: {:?}", contributor);
+                        continue;
+                    }
+
+                    // Extract signature
+                    let signature = match message.clone().payload {
+                        Some(Payload::Signature(signature)) => signature,
+                        _ => {
+                            info!("signature not found: {:?}", message.clone().payload);
+                            continue;
+                        }
+                    };
+                    let Ok(signature) = Sig::try_from(signature.clone()) else {
+                        info!("not a valid signature: {:?}", signature);
+                        continue;
+                    };
+                    let mut buf = Vec::with_capacity(message.encode_size());
+                    message.write(&mut buf);
+                    let Ok(payload) = validator.validate_and_return_expected_hash(&buf).await
+                    else {
+                        info!(
+                            "failed to validate payload for contributor: {:?}",
+                            contributor
+                        );
+                        continue;
+                    };
+                    // Verify signature from contributor using aggregate_verify with single public key
+                    if !aggregate_verify(std::slice::from_ref(&s), None, &payload, &signature) {
+                        info!("invalid signature from contributor: {:?}", contributor);
+                        continue;
+                    }
+
+                    // Insert signature
+                    signatures.insert(*contributor, signature);
+
+                    // Check if should aggregate
+                    if signatures.len() < threshold {
+                        info!(
+                            "current signatures aggregated: {:?}, needed: {:?}, continuing aggregation",
+                            signatures.len(),
+                            threshold
+                        );
+                        continue;
+                    }
+
+                    // Enough signatures, aggregate
+                    let mut participating = Vec::new();
+                    let mut participating_g1 = Vec::new();
+                    let mut sigs = Vec::new();
+                    for i in 0..contributors.len() {
```

Comment:
replaced with `for (i, contributor) in contributors.iter().enumerate() {` and remove clippy warning `#![allow(clippy::needless_range_loop)]`